vm-setup/roles/v1aX_integration_test/tasks/node_reuse.yml

```diff
@@ -128,48 +156,61 @@
     ignore_errors: yes
 
   - name: Wait until all "{{ NUMBER_OF_BMH_KCP }}" machines become running and updated with new "{{ UPGRADED_K8S_VERSION }}" k8s version.
-    shell: |
-        kubectl get machines -n "{{ NAMESPACE }}" -o json | jq -r '.items[] | select (.status.phase == "Running") | select(.spec.version == "{{ UPGRADED_K8S_VERSION }}") | .status.phase' | grep -c "Running"
-    environment:
-      KUBECONFIG: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
-    register: updated_machines_all
+    k8s_info:
+      api_version: cluster.x-k8s.io/{{ CAPI_VERSION }}
+      kind: machines
+      namespace: "{{ NAMESPACE }}"
+      kubeconfig: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
+    register: machines
     retries: 200
     delay: 20
     ignore_errors: yes
-    until: updated_machines_all.stdout|int == 3
+    until:
+      - machines is succeeded
+      - machines.resources | filter_phase("running") | selectattr('spec.version', 'match', UPGRADED_K8S_VERSION) | length == (NUMBER_OF_BMH_KCP | int)
+
+  - name: Get provisioned BMH after upgrade in KubeadmControlPlane node reuse test scenario.
+    k8s_info:
+      api_version: metal3.io/v1alpha1
+      kind: BareMetalHost
+      kubeconfig: "/tmp/kubeconfig-{{ CLUSTER_NAME }}.yaml"
+    register: saved_bmhs
```

Comment:
where is the BMH state check condition?

Reply:
I do not think there is a way to register only provisioned bmhs after filter so I am saving the whole bmhs set (including provisioned) and filter it in the next set fact would you like to change task name or use both tasks in block ?